File: bun/src/thread.ts
import * as yaml from 'js-yaml';
import * as path from 'path';

// Status constants with const assertions for type safety
export const TERMINAL_STATUSES = ['resolved', 'superseded', 'deferred', 'reject'] as const;
export const ACTIVE_STATUSES = ['idea', 'planning', 'active', 'blocked', 'paused'] as const;
export const ALL_STATUSES = [...ACTIVE_STATUSES, ...TERMINAL_STATUSES] as const;

// Derived types from the const arrays
export type TerminalStatus = typeof TERMINAL_STATUSES[number];
export type ActiveStatus = typeof ACTIVE_STATUSES[number];
export type Status = typeof ALL_STATUSES[number];

// Frontmatter interface
export interface Frontmatter {
  id: string;
  name: string;
  desc: string;
  status: string;
}

// Thread class
export class Thread {
  path: string;
  frontmatter: Frontmatter;
  content: string;
  bodyStart: number;

  constructor(filePath: string) {
    this.path = filePath;
    this.frontmatter = { id: '', name: '', desc: '', status: '' };
    this.content = '';
    this.bodyStart = 0;
  }

  static async parseAsync(filePath: string): Promise<Thread> {
    const t = new Thread(filePath);
    t.content = await Bun.file(filePath).text();
    t.parseFrontmatter();

    // Extract ID from filename if not in frontmatter
    if (!t.frontmatter.id) {
      t.frontmatter.id = extractIDFromPath(filePath);
    }

    return t;
  }

  // Synchronous parse for backwards compatibility
  static parse(filePath: string): Thread {
    const t = new Thread(filePath);
    // Use Bun.file with sync read via require('fs')
    const fs = require('fs');
    t.content = fs.readFileSync(filePath, 'utf-8');
    t.parseFrontmatter();

    // Extract ID from filename if not in frontmatter
    if (!t.frontmatter.id) {
      t.frontmatter.id = extractIDFromPath(filePath);
    }

    return t;
  }

  parseFrontmatter(): void {
    if (!this.content.startsWith('---\n')) {
      throw new Error('missing frontmatter delimiter');
    }

    const end = this.content.indexOf('\n---', 4);
    if (end === -1) {
      throw new Error('unclosed frontmatter');
    }

    const yamlContent = this.content.substring(4, end);
    this.bodyStart = end + 4; // skip opening ---, yaml, closing ---, and newline

    const parsed = yaml.load(yamlContent);

    // Validate YAML parsing result
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
      throw new Error('Invalid frontmatter format: expected object');
    }

    const fm = parsed as Record<string, unknown>;
    this.frontmatter = {
      id: typeof fm.id === 'string' ? fm.id : '',
      name: typeof fm.name === 'string' ? fm.name : '',
      desc: typeof fm.desc === 'string' ? fm.desc : '',
      status: typeof fm.status === 'string' ? fm.status : '',
    };
  }

  get id(): string {
    return this.frontmatter.id;
  }

  get name(): string {
    return this.frontmatter.name;
  }

  get status(): string {
    return this.frontmatter.status;
  }

  baseStatus(): string {
    return baseStatus(this.frontmatter.status);
  }

  body(): string {
    if (this.bodyStart >= this.content.length) {
      return '';
    }
    return this.content.substring(this.bodyStart);
  }

  setFrontmatterField(field: string, value: string): void {
    switch (field) {
      case 'id':
        this.frontmatter.id = value;
        break;
      case 'name':
        this.frontmatter.name = value;
        break;
      case 'desc':
        this.frontmatter.desc = value;
        break;
      case 'status':
        this.frontmatter.status = value;
        break;
      default:
        throw new Error(`unknown field: ${field}`);
    }
    this.rebuildContent();
  }

  rebuildContent(): void {
    let sb = '---\n';

    // Serialize frontmatter in specific order
    const fm: Record<string, string> = {};
    if (this.frontmatter.id) fm.id = this.frontmatter.id;
    if (this.frontmatter.name) fm.name = this.frontmatter.name;
    if (this.frontmatter.desc !== undefined) fm.desc = this.frontmatter.desc;
    if (this.frontmatter.status) fm.status = this.frontmatter.status;

    sb += yaml.dump(fm, { lineWidth: -1 });
    sb += '---\n';

    // Preserve body
    if (this.bodyStart < this.content.length) {
      sb += this.content.substring(this.bodyStart);
    }

    this.content = sb;
  }

  async writeAsync(): Promise<void> {
    await Bun.write(this.path, this.content);
  }

  // Synchronous write for backwards compatibility
  write(): void {
    const fs = require('fs');
    fs.writeFileSync(this.path, this.content);
  }

  relPath(ws: string): string {
    return path.relative(ws, this.path);
  }
}

// ID prefix regex
const idPrefixRe = /^([0-9a-f]{6})-/;

// Extract 6-char hex ID from filename
export function extractIDFromPath(filePath: string): string {
  const filename = path.basename(filePath);
  const nameWithoutExt = filename.replace(/\.md$/, '');

  const match = nameWithoutExt.match(idPrefixRe);
  if (match && match[1]) {
    return match[1];
  }
  return '';
}

// Extract human-readable name from filename
export function extractNameFromPath(filePath: string): string {
  const filename = path.basename(filePath);
  const nameWithoutExt = filename.replace(/\.md$/, '');

  const match = nameWithoutExt.match(idPrefixRe);
  if (match) {
    return nameWithoutExt.substring(7); // skip "abc123-"
  }
  return nameWithoutExt;
}

// Strip reason suffix from status
export function baseStatus(status: string): string {
  const idx = status.indexOf(' (');
  if (idx !== -1) {
    return status.substring(0, idx);
  }
  return status;
}

// Check if status is terminal
export function isTerminal(status: string): boolean {
  const base = baseStatus(status);
  return (TERMINAL_STATUSES as readonly string[]).includes(base);
}

// Check if status is valid
export function isValidStatus(status: string): boolean {
  const base = baseStatus(status);
  return (ALL_STATUSES as readonly string[]).includes(base);
}

File: ruby/lib/threads/thread.rb
# frozen_string_literal: true

require 'yaml'
require 'tempfile'
require 'fileutils'

module Threads
  # Custom exception classes
  class Error < StandardError; end
  class ThreadNotFound < Error; end
  class InvalidStatus < Error; end
  class WorkspaceError < Error; end
  class AmbiguousReference < Error; end
  class ParseError < Error; end
  class GitError < Error; end

  # Status constants
  TERMINAL_STATUSES = %w[resolved superseded deferred rejected].freeze
  ACTIVE_STATUSES = %w[idea planning active blocked paused].freeze
  ALL_STATUSES = (ACTIVE_STATUSES + TERMINAL_STATUSES).freeze

  # Validate status against allowed values
  def self.validate_status!(status)
    base = base_status(status)
    return if ALL_STATUSES.include?(base)

    raise InvalidStatus, "Invalid status: #{status}. Must be one of: #{ALL_STATUSES.join(', ')}"
  end

  # Check if status is terminal
  def self.terminal?(status)
    base = base_status(status)
    TERMINAL_STATUSES.include?(base)
  end

  # Check if status is valid
  def self.valid_status?(status)
    base = base_status(status)
    ALL_STATUSES.include?(base)
  end

  # Extract base status (strip reason suffix)
  def self.base_status(status)
    return '' if status.nil?

    idx = status.index(' (')
    idx ? status[0...idx] : status
  end

  # Thread represents a parsed thread file
  class Thread
    attr_accessor :path, :content, :frontmatter, :body_start

    def initialize(path)
      @path = path
      @content = ''
      @frontmatter = {}
      @body_start = 0
    end

    # Parse a thread file
    def self.parse(path)
      t = new(path)
      t.content = File.read(path)
      t.parse_frontmatter
      t
    end

    # Parse YAML frontmatter
    def parse_frontmatter
      unless @content.start_with?("---\n")
        raise ParseError, 'missing frontmatter delimiter'
      end

      # Find closing delimiter
      end_idx = @content.index("\n---", 4)
      raise ParseError, 'unclosed frontmatter' unless end_idx

      yaml_content = @content[4...end_idx]
      @body_start = end_idx + 5 # skip opening ---, yaml, closing ---, and newline

      begin
        @frontmatter = YAML.safe_load(yaml_content) || {}
      rescue Psych::SyntaxError => e
        raise ParseError, "invalid YAML frontmatter: #{e.message}"
      end

      # Extract ID from filename if not in frontmatter
      if id.nil? || id.to_s.empty?
        @frontmatter['id'] = Workspace.extract_id_from_path(@path)
      end
    end

    def id
      @frontmatter['id']
    end

    def name
      @frontmatter['name']
    end

    def status
      @frontmatter['status']
    end

    def desc
      @frontmatter['desc']
    end

    def base_status
      Threads.base_status(status)
    end

    def terminal?
      Threads.terminal?(status)
    end

    # Get body content (after frontmatter)
    def body
      return '' if @body_start >= @content.length

      @content[@body_start..]
    end

    # Set frontmatter field and rebuild content
    def set_field(field, value)
      @frontmatter[field] = value
      rebuild_content
    end

    # Rebuild content from frontmatter and body
    def rebuild_content
      body_content = body
      @content = "---\n"
      @content += YAML.dump(@frontmatter).sub(/^---\n?/, '')
      @content += "---\n"
      @body_start = @content.length
      @content += body_content if body_content && !body_content.empty?
    end

    # Write thread to disk atomically
    def write
      atomic_write(@path, @content)
    end

    # Relative path from workspace
    def rel_path(ws)
      @path.sub("#{ws}/", '')
    end

    private

    # Atomic file write using temp file and rename
    def atomic_write(path, content)
      dir = File.dirname(path)
      Tempfile.create(['thread', '.tmp'], dir) do |f|
        f.write(content)
        f.close
        FileUtils.mv(f.path, path)
      end
    rescue SystemCallError => e
      raise Threads::Error, "Failed to write #{path}: #{e.message}"
    end
  end
end
